gwctl/pkg/policymanager/manager.go

```diff
@@ -0,0 +1,345 @@
+	}
+
+	crds := &apiextensionsv1.CustomResourceDefinitionList{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredCRDs.UnstructuredContent(), crds); err != nil {
+		return []apiextensionsv1.CustomResourceDefinition{}, fmt.Errorf("failed to convert unstructured CRDs to structured: %v", err)
+	}
+
+	return crds.Items, nil
+}
+
+// fetchPolicies will fetch all policy resources corresponding to the CRDs
+// present in policyCRDs.
+func fetchPolicies(ctx context.Context, dc dynamic.Interface, policyCRDs map[PolicyCrdID]PolicyCRD) ([]unstructured.Unstructured, error) {
+	var result []unstructured.Unstructured
+
+	for _, policyCRD := range policyCRDs {
+		gvr := schema.GroupVersionResource{
+			Group:    policyCRD.crd.Spec.Group,
+			Version:  policyCRD.crd.Spec.Versions[0].Name,
+			Resource: policyCRD.crd.Spec.Names.Plural, // CRD Kinds directy map to the Resource.
+		}
+
+		var policies *unstructured.UnstructuredList
+		var err error
+		if policyCRD.IsClusterScoped() {
+			policies, err = dc.Resource(gvr).List(ctx, metav1.ListOptions{})
+		} else {
+			policies, err = dc.Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+		}
+		if err != nil {
+			return result, err
+		}
+
+		result = append(result, policies.Items...)
+	}
+
+	return result, nil
+}
+
+// PolicyCrdID has the structurued "<CRD Kind>.<CRD Group>"
+type PolicyCrdID string
+
+type PolicyCRD struct {
+	crd apiextensionsv1.CustomResourceDefinition
+}
+
+// ID returns a unique identifier for this PolicyCRD.
+func (p PolicyCRD) ID() PolicyCrdID {
+	return PolicyCrdID(p.crd.Spec.Names.Kind + "." + p.crd.Spec.Group)
+}
+
+// IsValid return true if the PolicyCRD satisfies requirements for qualifying as
+// a Gateway Policy CRD.
+func (p PolicyCRD) IsValid() bool {
+	return p.IsInherited() || p.IsDirect() || p.crd.GetLabels()[gatewayPolicyLabelKey] == "true"
+}
+
+func (p PolicyCRD) IsInherited() bool {
+	return p.crd.GetLabels()[gatewayPolicyLabelKey] == "inherited"
+}
+
+func (p PolicyCRD) IsDirect() bool {
+	return p.crd.GetLabels()[gatewayPolicyLabelKey] == "direct"
+}
+
+func (p PolicyCRD) CRD() *apiextensionsv1.CustomResourceDefinition {
+	return p.crd.DeepCopy()
+}
+
+// IsClusterScoped returns true if the CRD is cluster scoped. Such policies can
+// be used to target a cluster scoped resource like GatewayClass.
+func (p PolicyCRD) IsClusterScoped() bool {
+	return p.crd.Spec.Scope == apiextensionsv1.ClusterScoped
+}
+
+type Policy struct {
+	u         unstructured.Unstructured
+	targetRef ObjRef
+	// Indicates whether the policy is supposed to be "inherited" (as opposed to
+	// "direct").
+	inherited bool
+}
+
+type ObjRef struct {
+	Group     string `json:",omitempty"`
+	Kind      string `json:",omitempty"`
+	Name      string `json:",omitempty"`
+	Namespace string `json:",omitempty"`
+}
+
+func PolicyFromUnstrucutred(u unstructured.Unstructured, policyCRDs map[PolicyCrdID]PolicyCRD) (Policy, error) {
```

Comment:
```suggestion
func PolicyFromUnstructured(u unstructured.Unstructured, policyCRDs map[PolicyCrdID]PolicyCRD) (Policy, error) {
```

Reply:
Oh no sorry about that. Fixed. Wow how did you catch that amazing.